Notifications

mannykary opened the pull request:
### What does this pull request aim to achieve? 💯

Web and Email Notifications

Tickets resolved:
resolves #945
resolves #1406
resolves #1426
resolves #1507
resolves #1741
resolves #1427
resolves #1428
resolves #1478
resolves #1531
resolves #1737
resolves #1738
resolves #1739
resolves #1740
resolves #1824
resolves #1791
resolves #1792
resolves #1794
resolves #1796
resolves #1797
resolves #1799
resolves #1805
resolves #1832
resolves #1886 

### Testing Scope

**OUT OF SCOPE: Anything related to direct messages. Please post on #1480 if you notice an issue with this feature.**

1. Instant notifications for likes/comments/shares
1. Emails for instant notifications (if unread for more than 5 minutes). **NOTE: emails for DMs is still a WIP.**
1. Emails for daily/weekly/biweekly digest. For this, you will either need to wait a day/week/2 weeks to verify the digest email content, or reach out to @mannykary to force the digest emails to send as requested. The digest emails is a summary/aggregation of all notifications received over that time period, both read and unread.
1. Unsubscribe link in emails, and the unauthenticated unsubscribe page (i.e. the page you see after clicking the unsubscribe link). Verify that this page can be accessed even when not signed in (copy+paste the link to incognito browser session)
1. Notification preferences page when signed in
1. Notification preferences are respected so that it does not send email if certain setting is disabled
1. Multiple likes by the same person on the same post (i.e. like/unlike/like repeatedly) should result in only one notification for the receiver
1. Notification page is sync'ed with Unsubscribe page (modify one then access another, should see the changes applied)
1. Notification page and Unsubscribe page work layout work fine with Web and Mobile view

<!--### 🚨Before submitting this pull request🚨:-->

<!--_Please do **NOT** submit this PR if you have not done the following:_-->

- [x] I have checked that no one else is working on similar changes.
- [x] I have read the latest [**README**](README.md) and understand the development workflow.
- [x] The name of this branch is: **`feature/<branch_name>`**, or **`hotfix/<branch_name>`** if this is a hotfix, where `<branch_name>` briefly describes the issue(s) resolved and is prefixed with the issue number(s) (e.g. `1127-1130-update-git-branching-model`). The branch is created in a cloned version of this repo, **not a forked repo**.
- [x] This branch is rebased/merged with the **latest staging** (or the **latest production** if it is a hotfix).
- [x] There are no merge conflicts.
- [x] There are no console warnings and errors.
- [x] On the right hand side of this PR, I have linked the appropriate issue(s) under "Issues" and added a project progress under "Projects".
- [x] The title describes the issue(s) being addressed using format: **`<issue_numbers> - <brief_issue_description>`** (e.g. `#1127/#1130 - Update Git Branching Model`)

mannykary committed:
#1478 Microservices CI/CD for notifications backend (#1479)

- Add services folder for microservices
- Add notification backend skeleton code
- Add notification backend Terraform code
- Add notification backend Github Action workflows

mannykary committed:
Add AWS environment variables

mannykary committed:
Add Cloudwatch rule to invoke Lambda every 5 minutes

mannykary committed:
Remove terraform plan steps from CI/CD - can leak sensitive keys

mannykary committed:
Fix terraform init/deploy steps

mannykary committed:
Rename AWS variables - cannot use reserved keys

mannykary committed:
Increase Lambda timeout to 15 minutes

mannykary committed:
WIP Notification Lambda

mannykary committed:
Notification model (#1639)

mannykary committed:
Add Mongo and Mustache NPM packages

mannykary committed:
Add DateHelper to calculate dates easily

mannykary committed:
Add logic to DatabaseHelper

mannykary committed:
Add email templates

mannykary committed:
Add TemplateBuilder

mannykary committed:
Update Lambda handler and service

mannykary committed:
Fix unit test

mannykary committed:
Include notificationId in email template

mannykary committed:
Add setEmailSentAt to DatabaseHelper

mannykary committed:
Add retryWrites to Mongo connection string

mannykary committed:
Remove try/catch in Mailer.send

mannykary committed:
Update Service process function

mannykary committed:
Fix issues found during dev testing

mannykary committed:
Fix Lambda response body

mannykary committed:
Add unit tests

mannykary committed:
Temporarily stub out email send

mannykary committed:
Fix Mailer

mannykary committed:
Make instant unread interval configurable

mannykary committed:
Minor fixes to NotificationService

mannykary committed:
Update Cloudwatch rule to pass in frequency

mannykary committed:
Explicit check for DATABASE_RETRY_WRITES is "true"

mannykary committed:
Add more Lambda environment variables

mannykary committed:
Add Logger/Papertrail

mannykary committed:
Add Sentry error notifier

mannykary committed:
Add Sentry/Papertrail environment variables

mannykary committed:
Add FROM_EMAIL_ADDRESS env var

mannykary committed:
Add base URL environment variable

mannykary committed:
Use specific review build for base url for now

mannykary committed:
Fix template links to post

mannykary committed:
Fix staging/development base URLs

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

ddveloper committed:
update endpoints for notifyPrefs, add /unsubscribe

Zaydme committed:
push notifications

ddveloper committed:
add GET endpoint for /unsubscribe

ddveloper committed:
Merge branch 'feature/1637-notifications-web-backend' into feature/1427-notifications-unsubscribe

ddveloper committed:
Merge branch 'feature/1428' into feature/1427-notifications-unsubscribe

jessica-lau committed:
Add Fixture Files for Cypress Tests (#1674)

* added fixture files and updated sign in spec

* fixed cypress.json file

* added additional fixtures and updated tests

* added additional fixtures and updated tests

sujatha-m committed:
#1663 - Modified Quit filters button to Clear filter button (#1668)

* Modified Quit filters button to Clear filter button

* Moved the fix from en_GB.json to en_US.json

Zaydme committed:
(fix) user intercation with own post shouldn't trigger notifier

Zaydme committed:
Post share notifications

ddveloper committed:
follow the alphabeta order

ddveloper committed:
return limited information for /unsubscribe (unauthenticated endpoint)

ddveloper committed:
change 'post' to 'share'

ddveloper committed:
in case no notifyPreference when creating user, add default instance

jessica-lau committed:
Cypress e2e - Add cookie banner/logo POM files and refactor tests (#1706)

* added cookie banner and logo POM files and refactored e2e tests

* refactored additional tests

ddveloper committed:
undo non-relavant change

ddveloper committed:
add edit notification option in Profile editing

ddveloper committed:
complete notification editing page

ddveloper committed:
unsubscribe link (update done, render partially done)

wadehrarshpreet committed:
improvement: Chat UI fixes

wadehrarshpreet committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

wadehrarshpreet committed:
improvement: settings chat UI fix

ddveloper committed:
toggle off notification button when all false

ddveloper committed:
bug fix, was committed these by accident

emanuelquerty committed:
#614 - check email not verified (#1703)

* Fixes internal server error message on login with email not verified

* Makes the comment more concise

wadehrarshpreet committed:
improvement: alertbox action button UI fix

wadehrarshpreet committed:
improvement: long touch menu options chat

ddveloper committed:
unsubscribe page, link preference to UI

ddveloper committed:
these are not add by me ;( could be the formatter robot?

ddveloper committed:
add a success alert after update

Zaydme committed:
Revert '#1448 - Add search bar on help board with a search button' This reverts commit 0df102b186287aa9475d7a0c2adb213b3b92880a.

Zaydme committed:
fixed unread indicator position

Zaydme committed:
internationalization of messaging feature

Zaydme committed:
adjusted bubble border radius

mannykary committed:
Change prefix to service to prevent clobbered deploys

ddveloper committed:
remove jwt-simple, use jsonwebtoken instead

mannykary committed:
Add support for share notifications

mannykary committed:
Add TemplateBuilder test for comment

ddveloper committed:
Merge branch 'feature/1637-notifications-web-backend' into feature/1427-notifications-unsubscribe

mannykary committed:
Generate unsubscribe token

mannykary committed:
Combine notification preferences/unsubscribe link

mannykary committed:
Set base_url to Dawei's branch for now

mannykary committed:
Fix unsubscribe link in template

ddveloper committed:
update expireDate interface

Zaydme committed:
fixed typos en_US (messaging)

wadehrarshpreet committed:
fix: message ui fix left sidebar & other issue

wadehrarshpreet committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
Merge remote-tracking branch 'origin/feature/notifications' into feature/1637-notifications-web-backend

Zaydme committed:
Merge pull request #1677 from FightPandemics/feature/1427-notifications-unsubscribe

Feature/1427 notifications unsubscribe

wadehrarshpreet committed:
fix: message request issue fix

Zaydme committed:
ignored threads turn back into requests on new message

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
Merge branch 'feature/1637-notifications-web-backend' of https://github.com/FightPandemics/FightPandemics into feature/1637-notifications-web-backend

wadehrarshpreet committed:
improvement: confirm dialog added for ignore & block & minor ui fixes

wadehrarshpreet committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

ddveloper committed:
update notification setting menu words

wadehrarshpreet committed:
fix: unarchive screen & message modal ui fix(mobile)

mannykary committed:
Merge branch 'staging' into feature/notifications

mannykary committed:
Merge branch 'feature/notifications' into feature/1637-notifications-web-backend

mannykary committed:
Point review to feature/notifications branch build

mannykary committed:
Remove unused DateHelper import

mannykary committed:
Add ShareMedium enum model

mannykary committed:
Add share medium logic

mannykary committed:
Merge pull request #1638 from FightPandemics/feature/1637-notifications-web-backend

Web notifications

mannykary committed:
Merge pull request #1720 from FightPandemics/service/1426-notifications-lambda

#1426 Notifications Lambda

mannykary committed:
Fix review build base URL

mannykary committed:
Merge branch 'staging' into feature/notifications

mannykary committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
backend code cleanup

Zaydme committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
(Messaging code changes) Merge branch 'feature/message-modal/1457' into feature/notifications

Commit:
#1739 - new notification preference schema update (#1742)

mannykary committed:
Add digest email templates

mannykary committed:
Add more enum models

mannykary committed:
DatabaseHelper: add digest email queries

mannykary committed:
TemplateBuilder: build digest email templates

mannykary committed:
Notification model: change emailSentAt to object

Need to track each email type separately so we don't count the same
notifications multiple times in digest emails

mannykary committed:
More robust emailSentAt logic

mannykary committed:
Set emailSentAt to object in new notification

Zaydme committed:
Added commentText to Notification schema

Zaydme committed:
internationalization of web notifications

Zaydme committed:
i didn't add these, reverting.

Zaydme committed:
changed link of notifications dropdown settings icon

mannykary committed:
Use npm ci in CI/CD

This ensures repeatable, deterministic builds
https://stackoverflow.com/a/53325242

wadehrarshpreet committed:
improvement: updated profile page UI

Zaydme committed:
Added GTM for messaging

Zaydme committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

wadehrarshpreet committed:
fix: minor style fix

wadehrarshpreet committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

mannykary committed:
Merge pull request #1747 from FightPandemics/feature/1740-notif

Feature/1740 notif

mannykary committed:
Fix issues in DatabaseHelper

mannykary committed:
local.js - pass frequency as environment variable

mannykary committed:
Add daily/weekly/biweekly cron jobs

mannykary committed:
Run terraform fmt

mannykary committed:
Merge branch 'staging' into feature/notifications

# Conflicts:
#	client/src/locales/translations/en_GB.json

mannykary committed:
Add feature/notifications to review branch trigger

Zaydme committed:
deduplicate likes on a single post by the same user

Zaydme committed:
Small UI fixes

mannykary committed:
Merge pull request #1771 from FightPandemics/feature/1738-notifications-deduplicate

Notifications - deduplicate likes on a single post by the same user

mannykary committed:
Merge pull request #1770 from FightPandemics/feature/1737-notification-digest-emails

#1737 notification digest emails

mannykary committed:
Fix notification lambda unit tests

mannykary committed:
Fix Cloudwatch cron expressions

mannykary committed:
Show commenter name in digest emails

mannykary committed:
Show comment text in instant emails

mannykary committed:
Add extra padding to main div

mannykary committed:
DatabaseHelper: find unread direct messages

wadehrarshpreet committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

wadehrarshpreet committed:
fix: profile ui fix

wadehrarshpreet committed:
improvement: updated org profile view

mannykary committed:
Merge branch 'staging' into feature/notifications

# Conflicts:
#	client/src/components/Feed/FilterAccordion.js
#	client/src/templates/layouts/NavigationLayout.js

mannykary committed:
Fix botched merge for FilterAccordion.js

ddveloper committed:
adjust notification preference layout

Zaydme committed:
fix: newly signed up users not getting connected to WS

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/notifications

Zaydme committed:
Merge branch 'staging' into feature/notifications

Zaydme committed:
Added missing comma after merge conflict

Zaydme committed:
UI improvements: web notifications.

Zaydme committed:
Merge branch 'feature/notifications' of https://github.com/FightPandemics/FightPandemics into feature/notifications

Zaydme committed:
mobile: hide chat list and show current chat when clicking "View Message"

Zaydme committed:
Updated messaging models indexes

Zaydme committed:
fixed bug: user can send message to pending conversation from helpboard before accepting it.

Zaydme committed:
Message button on Poeple/Orgs cards

Zaydme committed:
fixed small bug caused by previous commit

ddveloper committed:
UI improvement: unsubscribe

ddveloper committed:
lost Notification link by merge, add it back

mannykary commented on client/src/locales/translations/en_GB.json:
```diff
@@ -832,7 +832,8 @@
     "tryAgainBtn": "Try Again"
   },
   "unsubscribe": {
-    "notifyPrefs": "Notification Preference",
```
@ddveloper Just FYI, we shouldn't be updating any other translation file other than `en_US.json` since Lokalise will automatically create a pull request to translate from `en_US.json` to the other languages. See https://github.com/FightPandemics/FightPandemics#internationalization

ddveloper replied:
Got it, thanks Manny, I will create PR next time 👍

ddveloper committed:
increase linkage font size

mannykary committed:
Merge pull request #1795 from FightPandemics/feature/notificationLinkFont

increase linkage font size

Commit:
fix cancel operation wording (#1800)

ddveloper committed:
save previous checks in localStorage

Zaydme committed:
#1794 - Notifications dropdown UI improvements (#1803)

Commit:
Merge pull request #1801 from FightPandemics/feature/1799-notifications-keepChecks

save previous checks in localStorage

wadehrarshpreet committed:
fix: feed color in myprofile

ddveloper committed:
extract notifyPrefs for both user and organisation

ddveloper committed:
add OrganisationNotification page

ddveloper committed:
add link in organisation profile page

mannykary committed:
Merge pull request #1804 from FightPandemics/feature/1797-notification-OrgSetting

Feature/1797 notification org setting

Commit:
keep checkbox values for Org-Notification-Preference (#1809)

mannykary committed:
Merge branch 'staging' into feature/notifications

mannykary committed:
Add indexes and emailSent field

mannykary committed:
Tie user object to participant in query

mannykary committed:
Set useUnifiedTopology to fix deprecation warning

Zaydme committed:
removed unused indexes

Zaydme committed:
Combined some duplicated socket events into one

Zaydme committed:
Fixed user status not updating in realtime

Zaydme committed:
messaging dates internationalizations (global moment.js internationalization)

Zaydme committed:
UI fixes and general UX improvements

Zaydme committed:
made dropdowns trigger on click only

Zaydme committed:
auto focus editing textarea

mannykary committed:
Add direct message email templates

mannykary committed:
Fix logic to query for unread DMs

mannykary committed:
Add message to EmailFrequency enum

mannykary committed:
TemplateBuilder: add DM template logic

mannykary committed:
Add message frequency Cloudwatch rule

mannykary committed:
Add message to set of valid frequencies

mannykary committed:
Remove TODO comment

mannykary committed:
Use MessageThreadStatus.PENDING enum

mannykary committed:
Refine emailSent flipping logic

mannykary committed:
Add postTextCopy to messageRequest template

mannykary committed:
Merge pull request #1822 from FightPandemics/feature/1741-update-message-thread-schemas

#1741 update message thread schemas

mannykary committed:
Merge branch 'staging' into feature/notifications

mannykary committed:
Merge branch 'feature/message-modal/1457' into feature/notifications

mannykary committed:
Fix digest email styling

Zaydme committed:
web notifications date i18n

mannykary committed:
Merge pull request #1825 from FightPandemics/feature/web-notifications-date-i18n

Web notifications date i18n

mannykary committed:
Merge pull request #1821 from FightPandemics/nodeploy/1741-dm-email-notifications

#1741 Email notifications for DMs

Zaydme committed:
Fix notification digest email styling - Merge pull request #1823 from FightPandemics/nodeploy/fix-digest-email-styling

Fix notification digest email styling

mannykary committed:
Use elemMatch; filter out messages older than 30m

- Need to use $elemMatch to match document in array field so that
all conditions are satisfied by the document in the array
- Move safety filter to messages query to avoid sending notifications
for old messages, since the thread lastAccess timestamp does not
indicate how old a message is

mannykary committed:
Merge pull request #1830 from FightPandemics/service/fix-dm-notifications

Fix DM email notifications

ddveloper committed:
use BaseUser which supports both Individual and Organisations for unsubscribing

Commit:
Merge pull request #1831 from FightPandemics/feature/notifications-unsubscribe-baseUser

use BaseUser which supports both Individual and Organisations for uns…

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Messaging as Org

Zaydme committed:
Fix mobile blank page after unarchiving from settings

Zaydme committed:
Fix unarchive button not visible

Zaydme committed:
Hide contextMenu from deleted messages

Zaydme committed:
Fix scrollbar style in messageModal

Zaydme committed:
Prevent mobile keyboard from bouncing (close/open)

Zaydme committed:
Fix "Help board" button on post page when post not found

Zaydme committed:
fix inbox header icon not showing correct number

Zaydme committed:
Now you can't block someone before they accept your request

Zaydme committed:
Fix messageModal button hover style

Zaydme committed:
organization profile UI fixes

Zaydme committed:
message modal disabled button styling

Zaydme committed:
update org avatar in thread participant ref

Zaydme committed:
fix merge conflicts

Zaydme committed:
add notifications dropdown and icon

Zaydme committed:
Updated notifier logic to work with actorId

Zaydme committed:
UI improvements + requested changes

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/notifications-with-act-as-org

Zaydme committed:
check if owner interacted with owned org posts

Zaydme committed:
UI fixes related to account switch

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/notifications-with-act-as-org

Zaydme committed:
change edit notifications link depending on actor.

mannykary committed:
Merge pull request #1867 from FightPandemics/feature/notifications-with-act-as-org

Notifications with act as org

mannykary committed:
Use noreply email for notifications

Zaydme committed:
merged staging into notifications

mannykary committed:
Merge pull request #1877 from FightPandemics/feature/notifications-merged-with-stg

Notifications merged with STG

Zaydme committed:
Merge branch 'staging' into feature/notifications

Zaydme committed:
Adds GTM ids to web notifications

mannykary committed:
Remove feature/notifications branch hook

mannykary committed:
Merge branch 'feature/notifications' of https://github.com/FightPandemics/FightPandemics into feature/notifications

Zaydme committed:
Adds GTM id for notifications settings icon

Zaydme committed:
Merge branch 'feature/notifications' of https://github.com/FightPandemics/FightPandemics into feature/notifications

Zaydme committed:
remove socketio logs from PROD

mannykary committed:
Notification emails: add link tracking

mannykary committed:
Merge pull request #1886 from FightPandemics/add-notification-email-link-tracking

#1876 Notification emails: add link tracking

Zaydme reviewed (approved):
Let's go!
